esmvalcore/_recipe.py

```diff
@@ -1374,6 +1454,32 @@ def _resolve_diagnostic_ancestors(self, tasks):
                         ancestors.extend(tasks[a] for a in ancestor_ids)
                     tasks[task_id].ancestors = ancestors
 
+    def _fill_wildcards(self, variable_group, preprocessor_output):
+        # To be generalised for other tags
+        var = preprocessor_output[variable_group][0]
+        index = var.get('recipe_dataset_index')
+        diagnostic = var.get('diagnostic')
+
+        datasets = self._raw_recipe.get('datasets')
+        timerange = var.get('timerange')
+        if timerange:
+            if datasets:
+                raw_timerange = (
+                    self._raw_recipe['datasets'][index]['timerange'])
+            else:
+                raw_timerange = (self._raw_recipe['diagnostics'][diagnostic]
+                                 ['additional_datasets'][index]['timerange'])
+
+            if timerange != raw_timerange:
+                if not self._updated_recipe:
+                    self._updated_recipe = deepcopy(self._raw_recipe)
+                if datasets:
+                    (self._updated_recipe['datasets'][index]['timerange']
+                     ) = timerange
+                else:
+                    (self._updated_recipe['diagnostics'][diagnostic]
+                     ['additional_datasets'][index]['timerange']) = timerange
+
```

Comment:
What about the case where `additional_datasets` is specified under a variable?

Reply:
Forgot about it. In any case, is there a better way to access nested dicts? This function does its job, but I am not too conviced it's the best way to go about it.

Reply:
You could write a function to do it and use that, that would be nicer.

Looking at this code in more detail, I think it doesn't work because the 'recipe_dataset_index' is computed by adding the 3 lists of datasets (at recipe, diagnostic, and variable level), so this will give the wrong position in the list if you do not take the offset into account. Would it be possible to implement this function without using the recipe_dataset_index? I only added that because some NCL diagnostics apparently depend on the ordering of the datasets in the recipe, so it is used to order the datasets before writing them to the NCL interface. I think it's not good practice to use numerical indices when there are easier to understand alternatives. It is very easy to make a mistake like this with numerical indices.

Reply:
I have been looking into it without much success. Since datasets can get expanded, it's quite difficult to map the preprocessor output to the datasets in the recipe.

Reply:
It looks like expansion and `recipe_dataset_index` do not work together very well, currently, all expanded datasets get the same value for `recipe_dataset_index`, so ordering them afterwards is not possible. I think this index should only be computed after expansion.

To get this to work with the expansion, I believe that you'll first need to expand, then find the dateranges for every single dataset, and then (optionally) unexpand again if this is still possible, because the files of the different datasets may not share the same daterange and/or start_year,end_year. This probably works best if you move the call to `_update_timerange` right after `variables = self._expand_tag(variables, 'sub_experiment')` on line 1242, because then you don't need to bother with the preprocessor output.

Reply:
Thanks for the tip, let's see if I can make it work with this.